src/commands.js

```diff
@@ -173,7 +175,243 @@ module.exports = {
+      // allow user.pubkey notation
+      if (key && key.length !== 64 && key.indexOf(".") >= 0) {
+        key = getFullKey(cabal, key.split(".")[1])
+      }
+      var reason = args.slice(2).join(' ')
+      if (args[0] === 'add') {
+        cabal.core.addMod(key, { channel, reason }, (err) => {
+        if (err) { res.error(err) }
+        else {
+          res.info(`${getPeerName(cabal, key)} has been added as a moderator`)
+          res.end()
+        }
+        })
+      } else if (args[0] === 'remove') {
+        cabal.core.removeMod(key, { channel, reason }, (err) => {
+        if (err) { res.error(err) }
+        else {
+          res.info(`${getPeerName(cabal, key)} has been removed from being a moderator`)
+          res.end()
+        }
+        })
+      } else if (args[0] == 'list') {
+        pump(cabal.core.moderation.listMods(channel), to.obj(write, end))
+      } else {
+        usage()
+      }
+      function write (row, enc, next) {
+        res.info(Object.assign({}, row, {
+          text: `${getPeerName(cabal, row.id)} [${row.role}]`
+        }))
+        next()
+      }
+      function end (next) {
+        res.end()
+        next()
+      }
+    }
+  },
+  admin: {
+    help: () => 'add or remove an admin cabal-wide',
+    call: (cabal, res, arg) => {
+      var args = (arg || '').split(/\s+/)
+      if (args[0].length === 0) {
+        res.info('usage: /admin (add|remove) KEY {REASON...}')
+        res.info('usage: /admin list')
+        return res.end()
+      }
+      var channel = '@' // experiment with only setting administrators cabal-wide
+      var key = args[1]
+      // allow user.pubkey notation
+      if (key && key.length !== 64 && key.indexOf(".") >= 0) {
+        key = getFullKey(cabal, key.split(".")[1])
+      }
+      var reason = args.slice(2).join(' ')
+      if (args[0] === 'add') {
+        cabal.core.addAdmin(key, { channel, reason }, (err) => {
+        if (err) { res.error(err) }
+        else {
+          res.info(`${getPeerName(cabal, key)} has been added as an administrator`)
+          res.end()
+        }
+        })
+      } else if (args[0] === 'remove') {
+        cabal.core.removeAdmin(key, { channel, reason }, (err) => {
+        if (err) { res.error(err) }
+        else {
+          res.info(`${getPeerName(cabal, key)} has been removed from being an administrator`)
+          res.end()
+        }
+        })
+      } else if (args[0] == 'list') {
+        pump(cabal.core.moderation.listAdmins(channel), to.obj(write, end))
+      } else {
+        res.error(usage)
+      }
+      function write (row, enc, next) {
+        res.info(Object.assign({}, row, {
+          text: `${getPeerName(cabal, row.id)}`
+        }))
+        next()
+      }
+      function end (next) {
+        res.end()
+        next()
+      }
+    }
+  },
+}
+
+function getFullKey (details, key) {
+  const keys = Object.keys(details.getUsers())
+  return keys.filter((k) => k.startsWith(key))[0]
```

Comment:
Could this be a problem if `filter` returns more than one key match? Wouldn't we risk doing a moderation op on the wrong user?

Reply:
@cblgh @substack Do either of you know this code better & can assess if there's an issue?

Reply:
Dug into this! Feels pretty solid now to me: 2aebcc2ba2eab3a0d0b28a711046b5bc80165453